### PR TITLE
Create word search algorithm

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Create a word search algorithm

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by CaedenPH** [comment]
@tianyizheng02 Pinging for review

File: other/word_search.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+    def __init__(self, words: list[str], width: int, height: int) -> None:
+        self.words = words
+        self.width = width
+        self.height = height
+
+        # Board matrix holding each letter
+        self.board: list[list[str | None]] = [[None] * width for _ in range(height)]
+
+    def insert_north(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Comment (tianyizheng02) [inline other/word_search.py:37]: It'd help to have the block comment explain what `rows` and `cols` are for

File: other/word_search.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+                # Only check to be made here is if there are existing letters
+                # above the column that will be overwritten
+                letters_above = [self.board[row - i][col] for i in range(word_length)]
+                if all(letter is None for letter in letters_above):
+                    # Successful, insert the word north
+                    for i in range(word_length):
+                        self.board[row - i][col] = word[i]
+                    return
+
+    def insert_north_east(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Comment (tianyizheng02) [inline other/word_search.py:69]:
```suggestion
    def insert_northeast(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Nitpick: `north_east` -> `northeast`

File: other/word_search.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_left = [self.board[row][col + i] for i in range(word_length)]
+                if all(letter is None for letter in letters_left):
+                    # Successful, insert the word north
+                    for i in range(word_length):
+                        self.board[row][col + i] = word[i]
+                    return
+
+    def insert_south_east(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Comment (tianyizheng02) [inline other/word_search.py:139]:
```suggestion
    def insert_southeast(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Nitpick: `south_east` -> `southeast`

File: other/word_search.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+                if word_length + col > self.width:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_diagonal_left = [
+                    self.board[row - i][col + i] for i in range(word_length)
+                ]
+                if all(letter is None for letter in letters_diagonal_left):
+                    # Successful, insert the word north
```
Comment (tianyizheng02) [inline other/word_search.py:102]:
```suggestion
                    # Successful, insert the word northeast
```

File: other/word_search.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+            for col in cols:
+                # Check if there is space to the right of the word
+                if word_length + col > self.width:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_left = [self.board[row][col + i] for i in range(word_length)]
+                if all(letter is None for letter in letters_left):
+                    # Successful, insert the word north
```
Comment (tianyizheng02) [inline other/word_search.py:134]:
```suggestion
                    # Successful, insert the word east
```

File: other/word_search.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        >>> ws.insert_south_east("at", [1, 0], [2, 1, 0])
+        >>> ws.board  # doctest: +NORMALIZE_WHITESPACE
+        [['c', None, None],
+        ['a', 'a', None],
+        [None, 't', 't']]
+        """
+        word_length = len(word)
+        # Attempt to insert the word into each row and when successful, exit
+        for row in rows:
+            # Check if there is space for the word above the row
```
Comment (tianyizheng02) [inline other/word_search.py:156]:
```suggestion
            # Check if there is space for the word below the row
```

File: other/word_search.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+            if word_length + row > self.height:
+                continue
+
+            # Attempt to insert the word into each column
+            for col in cols:
+                # Check if there is space to the right of the word as well as above
```
Comment (tianyizheng02) [inline other/word_search.py:162]:
```suggestion
                # Check if there is space to the right of the word as well as below
```

File: other/word_search.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+                if word_length + col > self.width:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_diagonal_left = [
+                    self.board[row + i][col + i] for i in range(word_length)
+                ]
+                if all(letter is None for letter in letters_diagonal_left):
+                    # Successful, insert the word north
```
Comment (tianyizheng02) [inline other/word_search.py:172]:
```suggestion
                    # Successful, insert the word southeast
```

File: other/word_search.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+                # Only check to be made here is if there are existing letters
+                # below the column that will be overwritten
+                letters_below = [self.board[row + i][col] for i in range(word_length)]
+                if all(letter is None for letter in letters_below):
+                    # Successful, insert the word south
+                    for i in range(word_length):
+                        self.board[row + i][col] = word[i]
+                    return
+
+    def insert_south_west(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Comment (tianyizheng02) [inline other/word_search.py:209]:
```suggestion
    def insert_southwest(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Nitpick: `south_west` -> `southwest`

File: other/word_search.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        >>> ws.insert_south_west("at", [1, 2], [2, 1, 0])
+        >>> ws.board  # doctest: +NORMALIZE_WHITESPACE
+        [[None, None, 'c'],
+        [None, 'a', 'a'],
+        ['t', 't', None]]
+        """
+        word_length = len(word)
+        # Attempt to insert the word into each row and when successful, exit
+        for row in rows:
+            # Check if there is space for the word above the row
```
Comment (tianyizheng02) [inline other/word_search.py:226]:
```suggestion
            # Check if there is space for the word below the row
```

File: other/word_search.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+            if word_length + row > self.height:
+                continue
+
+            # Attempt to insert the word into each column
+            for col in cols:
+                # Check if there is space to the right of the word as well as above
```
Comment (tianyizheng02) [inline other/word_search.py:232]:
```suggestion
                # Check if there is space to the left of the word as well as below
```

File: other/word_search.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+                if word_length > col + 1:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_diagonal_left = [
+                    self.board[row + i][col - i] for i in range(word_length)
+                ]
+                if all(letter is None for letter in letters_diagonal_left):
+                    # Successful, insert the word north
```
Comment (tianyizheng02) [inline other/word_search.py:242]:
```suggestion
                    # Successful, insert the word southwest
```

File: other/word_search.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+            for col in cols:
+                # Check if there is space to the left of the word
+                if word_length > col + 1:
+                    continue
+
+                # Check if there are existing letters
+                # to the left of the column that will be overwritten
+                letters_left = [self.board[row][col - i] for i in range(word_length)]
+                if all(letter is None for letter in letters_left):
+                    # Successful, insert the word north
```
Comment (tianyizheng02) [inline other/word_search.py:274]:
```suggestion
                    # Successful, insert the word west
```

File: other/word_search.py [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+                    for i in range(word_length):
+                        self.board[row][col - i] = word[i]
+                    return
+
+    def insert_north_west(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Comment (tianyizheng02) [inline other/word_search.py:279]:
```suggestion
    def insert_northwest(self, word: str, rows: list[int], cols: list[int]) -> None:
```
Nitpick: `north_west` -> `northwest`

File: other/word_search.py [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+        word_length = len(word)
+        # Attempt to insert the word into each row and when successful, exit
+        for row in rows:
+            # Check if there is space for the word above the row
+            if word_length > row + 1:
+                continue
+
+            # Attempt to insert the word into each column
+            for col in cols:
+                # Check if there is space to the right of the word as well as above
```
Comment (tianyizheng02) [inline other/word_search.py:302]:
```suggestion
                # Check if there is space to the left of the word as well as above
```

File: other/word_search.py [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+                if word_length > col + 1:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_diagonal_left = [
+                    self.board[row - i][col - i] for i in range(word_length)
+                ]
+                if all(letter is None for letter in letters_diagonal_left):
+                    # Successful, insert the word north
```
Comment (tianyizheng02) [inline other/word_search.py:312]:
```suggestion
                    # Successful, insert the word northwest
```

File: other/word_search.py [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+        )
+        for word in self.words:
+            # Shuffle the row order and column order that is used when brute forcing
+            # the insertion of the word
+            rows, cols = list(range(self.height)), list(range(self.width))
+            shuffle(rows)
+            shuffle(cols)
+
+            # Insert the word via the direction
+            choice(directions)(word, rows, cols)
```
Comment (tianyizheng02) [inline other/word_search.py:344]: Isn't it possible for the `insert_...` functions to fail to insert a word? Can we ensure that the word will actually be inserted?

File: other/word_search.py [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+
+    result = ""
+    for row in range(len(board)):
+        for col in range(len(board[0])):
+            character = "#"
+            if (letter := board[row][col]) is not None:
+                character = letter
+            # Empty char, so add a fake char
+            elif add_fake_chars:
+                character = chr(randint(97, 122))
```
Comment (tianyizheng02) [inline other/word_search.py:384]: Can we have input words restricted to just lowercase English letters? If you're just padding out the word search with random English lowercase letters and the user chooses words that use other characters (uppercase English letters, etc) then the inserted words would stand out.

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: other/word_search.py (hunk 18 above)
Comment (CaedenPH) [reply other/word_search.py:384]: We can limit to only lowercase english letters 👍

File: other/word_search.py (hunk 17 above)
Comment (CaedenPH) [reply other/word_search.py:344]: You are correct, the function will fail if there is not enough space on the board to insert the word. I will refactor the logic of the insertions to return whether or not the insertion was successful or not, and then raise and error if the insertion was unsuccessful

**Comment by tianyizheng02** [comment]
@CaedenPH pre-commit and build are failing due to incorrect function names, please fix:
```
other/word_search.py:329: error: "WordSearch" has no attribute "insert_north_east"; maybe "insert_northeast", "insert_northwest", or "insert_southeast"?  [attr-defined]
other/word_search.py:331: error: "WordSearch" has no attribute "insert_south_east"; maybe "insert_southeast", "insert_southwest", or "insert_northeast"?  [attr-defined]
other/word_search.py:333: error: "WordSearch" has no attribute "insert_south_west"; maybe "insert_southwest", "insert_southeast", or "insert_northwest"?  [attr-defined]
```

**Review by tianyizheng02 — APPROVED** [review]

**Pull request merged**